### PR TITLE
Latent distribution test size

### LINKED ISSUE #222: Investigate when it is safe to use `LatentDistributionTest` on differently sized graphs
Variance of latent position estimates scales with size of graph, this may cause graphs of different sizes to appear different to LDT just due to this difference in variance

### LINKED ISSUE #358: latent distribution test on directed inputs
currently, latent distribution test does not support directed inputs. although there are lines that collapse together two embeddings
https://github.com/neurodata/graspy/blob/429728e454c3e1656cf213000e975b5c42d1d8ba/graspy/inference/latent_distribution_test.py#L104-L106
we do not know whether the behavior will be as expected. we need to 

- [ ] run simulations that verify that the behavior of the tests are as expected on directed inputs (i.e. they are valid and consistent)
- [x] stop throwing a directed input error
- [x] write tests and etc.

### LINKED ISSUE #363: hsic works with kernels, not distances
unlike all other tests used as backends of the latent distribution test (via hyppo), HSic works on kernels, not distances (it is actually almost identical to DCorr, except that it expects a kernel function, not a distance, and hence flips the sign on the test statistic). we currently do nothing about that. some options to consider are:

- removing it completely (not a big loss, since, once again, DCorr does a similar thing)
- leaving it as is, but expecting a user to provide a callable metric. (if they try to provide "euclidean" - raise an exception)
- wrapping the currently used dissimilarity metrics into similarity measures (very obvious how to do it for the gaussian kernel. less so obvious for l1/l2, but obviously possible (for example by subtracting all from the max or doing 1/. )
- some sort of a combination of the two right above: for example, expecting a callable metric, but using a gaussian kernel if unspecified.

### LINKED ISSUE #370: returning test statistic in LDT
some practitioners (read: Vince, cep) only care about the test statistic and not the p-value. obviously one can still extract it if they perform the full test. however, that wastes time and resources. one can set the number of iterations to 1 to minimize that, but we can still do less. i propose to allow the number of permutations to be set to 0 (hyppo allows that, so really it is just a change in argument check). i am happy to do this, but:

this brings up the following questions: what should be happening to the fit_predict in that case? should it return the test statistic instead? or the p-value of 1? or NaN? should we be raising warnings?

and on a larger scale: should we really have this API? should fit predict return p-value, or a tuple of a p-value and a test statistic, like many other tests in python? furthremore, should it really be a class? once again, most tests in python that i have seen (scipy, statsmodels) are functions, not classes.

### PR DESCRIPTION
**Pull request by alyakin314**
#### Reference Issues/PRs
- Fixes #222 
- Fixes #358  
- Fixes #363 
- Fixes #370  (the 0 permutations part; for change in api see #375)

#### What does this implement/fix? Explain your changes.
For #222:
- introduced the sampling from the asymptotic distribution of the ase correction
- added corresponding unit tests

For #358:
- No longer raises an exception with directed inputs
- Added corresponding tests

For #363:
- added the capability to use sklearn kernels (for hsic test, which is same as mmd) 
- raising warnings when distance used with hsic or kernel used with hsic
- added corresponding unit tests

For #370:
- No longer raises an exception with 0 bootstraps
- fit_predict left unchanged (returns a nan if used with 0 bootstraps)

Miscellaneous:
- removed gaussian medial kernel implementation and instead importing it from hyppo
  - avoid code duplication 
  - that version is parallelizable
- the default changed from "dcorr" + "euclidean" to "hsic" + "gaussian". reasoning behind the change:
  1. in theory, dcorr and hsic are almost the same, except that one uses distances and the other uses kernels (they are in fact, equivalent for some choices of distances and kernels, as per distances and kernels EE paper). in practice "dcorr" + "euclidean" and "hsic"+"gaussian" either perform very similarly, or "hsic" has an advantage (see: https://mgc.neurodata.io/tutorials/independence/indep_power.html ) .
  2. Minh's nonpar paper, which proves the convergence uses Gretton's MMD (which is same as "use k-sample transform + hsic"), rather than Energy (which  is same as "use k-sample transform + dcorr"). i wouldn't be surprised if it is possible to prove a similar thing for Energy as well, but no one has yet.
  3. the test that we used to have before adding hyppo backend (MMD) is the same as "hsic" + "gaussian". obviously because of implementation differences, results of old sims won't be reproduced exactly, even when seeded, but this is the closest possible choice to be "backward-compatible"
  4. all my simulations for the size correction paper were performed with MMD. in theory, they should transfer to, say, dcorr and maybe other tests, but i am most confident in this specific choice.
- added an option to use workers=-1 to specify all possible threads (consistent with hyppo and sklearn)
- passing workers to sklearn's distances and kernels

#### Any other comments?
Does not address #177

### CODE REVIEW

**Comment by alyakin314** [comment]
self-reminders:

- [ ] test passing distances to hsic and kernels to other tests
- [ ] test having graphs of different size
- [ ] tutorial notebook

**Comment by alyakin314** [comment]
self-reminders:

- [ ] run a simulation to ensure that it is working
- [ ] tutorial notebook
- [ ] parallelize sampling

**Comment by alyakin314** [comment]
broke the ci because hyppo currently does not handle the gaussian kernel + k-sample + different sizes. opened a corresponding pr:

https://github.com/neurodata/hyppo/pull/93

**Comment by alyakin314** [comment]
ci still breaks because hyppo's version in pypi does not include my most recent pr

self-reminders:

- [x] run a simulation to ensure that it is working
- [ ] tutorial notebook

File: graspy/inference/latent_distribution_test.py [hunk 1]
```diff
@@ -174,6 +238,54 @@ def _embed(self, A1, A2):
...
 
+    def _estimate_correction_variances(self, X_hat, Y_hat, pooled=False):
+        # TODO it is unclear whether using pooled estimator provides more or
+        # less power. this should be investigated. should not matter under null.
+        N, d_X = X_hat.shape
+        M, d_Y = Y_hat.shape
+        if N == M:
+            X_sigmas = np.zeros((N, d_X, d_X))
+            Y_sigmas = np.zeros((M, d_Y, d_Y))
+        elif N > M:
```
Comment (bdpedigo) [inline graspy/inference/latent_distribution_test.py:249]: is it possible to write in such a way that you always operate on the bigger one if not pooled, as opposed to having the two branches of this if/else be basically the same w/ args flipped?

Comment (bdpedigo) [reply graspy/inference/latent_distribution_test.py:249]: i.e. X could always be assigned the bigger graph if not equal, or something like that

File: graspy/inference/latent_distribution_test.py [hunk 2]
```diff
@@ -174,6 +238,54 @@ def _embed(self, A1, A2):
...
+            return X, Y
+        elif n > m:
+            X_sigmas, _ = self._estimate_correction_variances(X, Y)
+            X_sampled = np.zeros(X.shape)
+            for i in range(n):
+                X_sampled[i, :] = X[i, :] + stats.multivariate_normal.rvs(
+                    cov=X_sigmas[i]
+                )
+            return X_sampled, Y
+        else:
```
Comment (bdpedigo) [inline graspy/inference/latent_distribution_test.py:280]: same comment about redundant if/else as above

File: tests/test_latentdistributiontest.py [hunk 3]
```diff
@@ -86,49 +96,51 @@ def test_directed_inputs(self):
...
-                test, "gaussian", n_components=2, n_bootstraps=50
-            )
-            ldt_alt = LatentDistributionTest(
-                test, "gaussian", n_components=2, n_bootstraps=50
-            )
-            p_null = ldt_null.fit_predict(A1, A2)
-            p_alt = ldt_alt.fit_predict(A1, A3)
-            self.assertTrue(p_null > 0.05)
-            self.assertTrue(p_alt <= 0.05)
+    def test_different_sizes(self):
```
Comment (bdpedigo) [inline tests/test_latentdistributiontest.py:121]: maybe also want a test for n != m but null is false? just to show power can be greater than 0

File: graspy/inference/latent_distribution_test.py [hunk 4]
```diff
@@ -45,24 +54,39 @@ class LatentDistributionTest(BaseInference):
...
-        :class:`hyppo.ksample.KSample` for more information.
+        Number of bootstrap iterations for the backend hypothesis test.
+        See :class:`hyppo.ksample.KSample` for more information.
 
     workers : int, optional (default=1)
         Number of workers to use. If more than 1, parallelizes the code.
+        Supply -1 to use all cores available to the Process.
+
+    size_correction: bool (default=True)
+        The size degrades in validity as the sizes of two graphs diverge from
```
Comment (bdpedigo) [inline graspy/inference/latent_distribution_test.py:83]: replace "size" w/ "test" i think

File: tests/test_latentdistributiontest.py (hunk 3 above)
Comment (alyakin314) [reply tests/test_latentdistributiontest.py:121]: good point.

File: graspy/inference/latent_distribution_test.py (hunk 4 above)
Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py:83]: hmmm... i don't feel too strongly about having this specific name, but i do not like "test_correction" too much, because it implies that it isn't correct without this flag. which is true, but only when graphs don't have the same number of vertices.

i can't remember whether i meant "size" as the number of vertices, or "size" as the thing that's less than alpha under the null. i feel like it was the former, but cep has convinced me that the number of vertices of the graph should be called "order of the graph" and not the "size of the graph", which is the number of edges, so maybe "order_correction" is more appropriate. thoughts?

File: graspy/inference/latent_distribution_test.py (hunk 1 above)
Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py:249]: i had internal debate about between elegance and readability when writing this. having two branches makes the symmetry more apparent, whereas abusing the order of the graphs is more concise. i feel like the two are about the same in terms of a probability of introducing a bug, since one has twice more code to mess up in, and the other has the ordering to keep track of. i went with the readability, but it was a 55-45 call, and since you feel the other way, i am happy to change it.

File: graspy/inference/latent_distribution_test.py (hunk 2 above)
Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py:280]: aaaand same as above

File: graspy/inference/latent_distribution_test.py (hunk 4 above)
Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py]: -_- didn't realize you were talking about the description... my bad

Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py]: will still change to "order_correction"

Comment (bdpedigo) [reply graspy/inference/latent_distribution_test.py]: yeah my comment was just on the description. but, i don't think "order" makes a ton of sense for n_verts to the average user. i was fine with `size_correction`. But you think "size" sounds like size of the test to a stats person?

Comment (bdpedigo) [reply graspy/inference/latent_distribution_test.py]: idk i always think of size as n_verts and not n_edges. maybe there's a third term that can make everyone happy tho

Comment (bdpedigo) [reply graspy/inference/latent_distribution_test.py]: still not loving the word "order" from a software perspective. any other terms we can think of?

**Comment by bdpedigo** [comment]
> ci still breaks because of hyppo versions (tests pass locally). i can change the requirements to pass it.

If you are saying this only works with a certain version of hyppo, then yeah please update requirements. Error I see is that one of your new tests is saying reject the null when null is true, that is related to hyppo somehow?

File: graspy/inference/latent_distribution_test.py [hunk 5]
```diff
@@ -125,37 +153,75 @@ def __init__(
...
+        elif n_bootstraps < 0:
+            msg = "{} is invalid number of bootstraps, must be positive"
             raise ValueError(msg.format(n_bootstraps))
 
         if not isinstance(workers, int):
             msg = "workers must be an int, not {}".format(type(workers))
             raise TypeError(msg)
-        elif workers <= 0:
-            msg = "{} is invalid number of workers, must be greater than 0"
+        elif workers == 0 or workers < 1:
```
Comment (bdpedigo) [inline graspy/inference/latent_distribution_test.py:163]: how does this allow workers to be -1? also, don't some systems allow -2 for all cores but one? not sure what the backend is here ultimately but sometimes that's a thing

File: graspy/inference/latent_distribution_test.py [hunk 6]
```diff
@@ -125,37 +153,75 @@ def __init__(
...
+                        "recomended to use either a hisc as a test or one of "
+                        f"the distances: {_VALID_DISTANCES} as a metric."
+                    )
+                    warnings.warn(msg, UserWarning)
 
                 def metric_func(X, Y=None, metric=metric, workers=None):
-                    return pairwise_distances(X, Y, metric=metric)
+                    return pairwise_kernels(X, Y, metric=metric, n_jobs=workers)
 
         self.test = KSample(test, compute_distance=metric_func)
```
Comment (bdpedigo) [inline graspy/inference/latent_distribution_test.py:216]: should probably document this (i know was not changed in this PR). Also I guess it's weird that we are overwriting an attribute "test" with an object

File: graspy/inference/latent_distribution_test.py [hunk 7]
```diff
@@ -170,10 +236,46 @@ def _embed(self, A1, A2):
...
-            raise ValueError("Input graphs do not have same directedness")
+            msg = (
+                "input graphs do not have same directedness. "
+                "consider symmetrizing the directed graph."
+            )
+            raise ValueError(msg)
 
         return X1_hat, X2_hat
 
+    def _sample_modified_ase(self, X, Y, pooled=False):
```
Comment (bdpedigo) [inline graspy/inference/latent_distribution_test.py:247]: pooled never gets used, then? since not a kwarg for the class? not sure whether to have this as a kwarg but if not, what is it doing?

File: graspy/inference/latent_distribution_test.py (hunk 4 above)
Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py]: i mostly don't want to annoy graph people who are used to the convention stated in wikipedia (see email i forwarded to you). the size_correction was intuitive to me as well, personally.

File: graspy/inference/latent_distribution_test.py (hunk 5 above)
Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py:163]: it doesn't.

this is me being sloppy. meant to be
`elif workers == 0 or workers < -1`

thanks for the catch.

File: graspy/inference/latent_distribution_test.py (hunk 6 above)
Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py:216]: just to clarify, you are saying to document the .test?

File: graspy/inference/latent_distribution_test.py (hunk 4 above)
Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py]: how much do you hate "correction"? i know i was initially opposed to not being clear about the effects only being present in the networks of different number of vertices, but this grew on me.

Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py]: alternatives:
- sampling
- variance_correction

Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py]: i take "sampling" back. i actually don't like it due to possible confusion with the permutations.

**Comment by alyakin314** [comment]
@bdpedigo ci works with the version of hyppo currently on the master, but breaks with the version of hyppo currently on pypi, because it has a bug in the gaussian kernel with an adaptive bandwidth. this bug breaks the test that uses graphs of different size and gaussian kernel.

passing the ci requires to set the requirements to a github link, which isn't very nice, but @sampan501 promised to iterate the version some time soon.


(summary of the bug if you care: if substantially many of the samples are equal to each other, it selects the bandwidth of 0, because the median between them is, well, zero. then there is division by 0 and it's downhill from there. now consider k-sample transform + one sample bigger than each other. then most of the data points in the auxiliary label vector are equal to each other, so you run into the aforementioned bug. i fixed it by setting the bandwidth to 1 whenever it becomes zero, and this really only happens with k-sample anyway.)

**Comment by bdpedigo** [comment]
> @bdpedigo ci works with the version of hyppo currently on the master, but breaks with the version of hyppo currently on pypi, because it has a bug in the gaussian kernel with an adaptive bandwidth. this bug breaks the test that uses graphs of different size and gaussian kernel.
> 
> passing the ci requires to set the requirements to a github link, which isn't very nice, but @sampan501 promised to iterate the version some time soon.
> 
> (summary of the bug if you care: if substantially many of the samples are equal to each other, it selects the bandwidth of 0, because the median between them is, well, zero. then there is division by 0 and it's downhill from there. now consider k-sample transform + one sample bigger than each other. then most of the data points in the auxiliary label vector are equal to each other, so you run into the aforementioned bug. i fixed it by setting the bandwidth to 1 whenever it becomes zero, and this really only happens with k-sample anyway.)

I see. Well, in short I wouldn't want to merge till that gets put on PyPI and we specify hyppo >= whatever that release will be. Agree that making it a github link is not ideal. @sampan501 just curious, how long do you think that would be? If it's < a week or two, let's just wait for that?

File: graspy/inference/latent_distribution_test.py (hunk 4 above)
Comment (bdpedigo) [reply graspy/inference/latent_distribution_test.py]: lol https://www.merriam-webster.com/dictionary/incommensurable

File: graspy/inference/latent_distribution_test.py (hunk 6 above)
Comment (bdpedigo) [reply graspy/inference/latent_distribution_test.py:216]: well im saying it's odd that normally one would be able to pass in parameters, and then in sklearn world, can always access the values of those parameters by doing object.parameter. In this case that's not true. I won't worry about it for now since we are already talking about what to do with the API

File: graspy/inference/latent_distribution_test.py (hunk 5 above)
Comment (bdpedigo) [reply graspy/inference/latent_distribution_test.py]: as far as i can tell this ultimately lands [here](https://github.com/neurodata/hyppo/blob/06ef0fc0bc450b60cf8cab07f1923e3df0d613d9/hyppo/_utils.py#L139)
if true, joblib will take ints less than -1, e.g. -2 means use all cores but one (which I use often). hyppo never even checks this value. I feel like maybe we just make sure it's an int for now?

**Comment by sampan501** [comment]
Actually, I'll be releasing it today! Just building and releasing/updating pypi

**Comment by alyakin314** [comment]
great, thanks @sampan501 !

File: graspy/inference/latent_distribution_test.py (hunk 4 above)
Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py]: i'm reverting it back to size_correction, and keeping it as True by default. we can eventually just phase out uncorrected test completely, as cep suggested, but for now i think there is no harm with leaving it as an option. it would be useful at least for me to have if i need any additional simulations for my paper, and maybe someone wants to reproduce old simulations.

File: graspy/inference/latent_distribution_test.py (hunk 7 above)
Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py:243]: you are correct that it is impossible to set, unless one goes into source code. i had it there because i was trying both, but has no conclusive evidence that using a pooled one has any benefit.

i predominantly decided to leave it in the code so that it would be easier to whoever decides to extensively study the behavior when using and not using it, similarly to how we had code for directed that's never getting used.

i don't mind allowing to set it as a class kwarg. the primary reason why i didn't is because i wanted to avoid confusion for users. there is nothing in the literature on using this, other than a line in my manuscript that says that hypothetically one can do so.

i am happy with either of the following three options:
- leave it as is (or leave it in commented out. which, for practical reasons is equivalent)
- make it a class kwarg
- remove it completely

File: graspy/inference/latent_distribution_test.py (hunk 5 above)
Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py]: fine with me.

Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py]: resolved.

File: graspy/inference/latent_distribution_test.py (hunk 6 above)
Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py:209]: ah... yeah, i understood you. yes, a way to resolve this would be to instantiate the hyppo object in fit rather than in init. happy not to worry about it due to the coming api change.

**Comment by alyakin314** [comment]
@bdpedigo ci passes. i believe the pooled thing is the only concern left?

File: graspy/inference/latent_distribution_test.py (hunk 4 above)
Comment (bdpedigo) [reply graspy/inference/latent_distribution_test.py]: agree and agree

File: graspy/inference/latent_distribution_test.py (hunk 7 above)
Comment (bdpedigo) [reply graspy/inference/latent_distribution_test.py:236]: my vote is make it a class kwarg, give an explanation in the docs, explain the default, and say that the not default way is untested/experimental

Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py:236]: sounds fair to me. doing dis now.

Comment (alyakin314) [reply graspy/inference/latent_distribution_test.py:254]: fixed. feel free to change the wording.
